test/common/index.js

```diff
@@ -703,30 +715,50 @@ exports.expectsError = function expectsError(fn, settings, exact) {
         typeName = Object.getPrototypeOf(error.constructor).name;
       }
       assert.strictEqual(typeName, type.name);
+      // Remove 'type'
+      keys.splice(keys.indexOf('type'), 1);
     }
-    if ('message' in settings) {
-      const message = settings.message;
-      if (typeof message === 'string') {
-        assert.strictEqual(error.message, message);
-      } else {
-        assert(message.test(error.message),
-               `${error.message} does not match ${message}`);
-      }
+
+    if ('message' in settings && typeof settings.message === 'object') {
+      assert(settings.message.test(error.message),
+             `${error.message} does not match ${settings.message}`);
+      // Remove 'message'
+      keys.splice(keys.indexOf('message'), 1);
     }
 
-    // Check all error properties.
-    const keys = Object.keys(settings);
     for (const key of keys) {
-      if (key === 'message' || key === 'type')
-        continue;
-      const actual = error[key];
-      const expected = settings[key];
-      assert.strictEqual(actual, expected,
-                         `${key}: expected ${expected}, not ${actual}`);
+      if (!Object.is(error[key], settings[key])) {
+        // Create placeholder objects to create a nice output.
+        const a = new Comparison(error, keys);
+        const b = new Comparison(settings, keys);
+
+        const tmpLimit = Error.stackTraceLimit;
+        Error.stackTraceLimit = 0;
+        const err = new assert.AssertionError({
+          actual: a,
+          expected: b,
+          operator: 'strictEqual',
+          stackStartFn: assert.throws,
+          errorDiff: 2
+        });
+        Error.stackTraceLimit = tmpLimit;
+
+        throw new assert.AssertionError({
+          actual: error,
+          expected: settings,
+          operator: 'common.expectsError',
+          message: err.message
+        });
+      }
+
     }
     return true;
   }
   if (fn) {
+    console.error(
+      'Please use `assert.throws()` instead of `common.expectsError()`. ' +
+      'To verify the error type, compare the error name.'
+    );
```

Comment:
I'm actually -1 on this bit. Adding `common.expectsError()` was deliberate to differentiate between checking ordinary JavaScript throws and uses of `internal/errors`. If the output for `assert.throws()` is better, then the better thing to do is improve the output for `common.expectsError()` accordingly.

Reply:
I am -1 as well. It's fairly easy to change `common.expectsError()` to meet whatever we need, but changing `assert.throws` comes with much more implications - IIRC `assert` should've not been exposed to users for the same reason.

Reply:
I am not sure I can follow. I can not remember the original implementation but the overloading that triggers `assert.throws` is implemented by me and that was done because `assert.throws` did not have the same capabilities as `common.expectsError` at that point of time. Since then I added support for objects in `assert.throws` and improved the error output significantly.

Right now there is still a valid use case for `common.expectsError`: in case it is used as callback function. But in all other cases I believe it is better to use `assert.throws` due to multiple reasons and it is also much more intuitive for new users. And the warning is there for exactly that use case: use `assert.throws` directly if possible.

Reply:
@joyeecheung as far as I know there were multiple discussions about this (in a general term). I believe code should never be "locked" as it is important to further improve things. It was the reason that the module became a wart in Node.js because people used it and stumbled upon deficiencies of it. Now the user experience is actually quite nice and people can use it out of my perspective. Even though there are of course more things that can be improved.

As far as I see it, you are against adding this comment. Are you against the PR in general? I have no strong feelings besides believing that we should use `assert.throws` and not `common.expectsError` but I will go with what ever is the opinion of others here.

Reply:
@BridgeAR I love the work you did on `assert.throws()` when I used them in our own tests, but I am still skeptical about actually changing the public assert significantly, given that it used to be locked - we did promise our users that it won't receive such a drastic change but we are now breaking the promise. Had I notice the changes to assert I would suggest to keep the changes in our internal test utilities for a while and have some discussion about this before actually exposing them to users, sorry.

BTW this is thread about removing the `locked` stability index. https://github.com/nodejs/node/issues/11200

Reply:
Another point about having a different implementation: that is somewhat correct but in that case we normally have something individual anyway. And it was partially hard to improve `common.expectsError`... even the benefit that it has now (comparing the `type` properly`) got a lot of mixed reactions first. So there was only one other benefit besides that: it is able to handle regular expressions for error messages.

Reply:
@BridgeAR I am not against this PR since it's about changing `common.expectsError`, but I am generally against encouraging changes to/usage of `assert` over that of our own internal test utilities, it should be the other way around, in my opinion. I am perfectly fine with fixing bugs in assert, or even adding support to new ECMAScript types, but I am skeptical about adding new, sometimes complex features to the public API, especially when those features don't have any spec or convention to adhere to, since more code usually means more bugs. At the very least, changes should live in our internal utilities for a while before actually being moved to the public API.

Reply:
@joyeecheung I am not aware of any bugs in `assert` but I am very happy to address such immediately, when they come up :-). It confuses me though, that implementing new features in `assert` has to be harder than implementing new features somewhere else in the code base?

Reply:
@BridgeAR https://github.com/nodejs/node/issues/19106 and https://github.com/nodejs/node/pull/19221 are still open. To be honest I don't really think there are good solutions to any of them, even the solutions I proposed are opinionated and users will always end up having to change their code to make it look right again in certain situations. It's not that implementing new features in public API is harder, but it's more disruptive to have opinionated, complex changes in the public API.

Reply:
I opened #19467 as a start for #19106. I plan on addressing that soon after :-). #19221 is something I would also like to have. Have you tried the suggestions?

Reply:
@BridgeAR The thing is, the more feature that assert receives, the more weird the whole things seem weird to me. As a user, I would not expect a good diff and color output from the built in assert module, at most I would just expect that it provides something that empowers user land to built such features on top of (e.g. properties on the Error object). I also don't see any good way out with the EOL issue to be honest, `.gitattributes` should be improved of course, but fixing it to make the assert work seems odd to me. We can't expect users all have proper `.gitattributes` set, or use git to deploy their running code at all.

Reply:
Hm, I do not see this as a issue. It is all well documented in the docs. So I think it depends on when the user read the docs the last time. My main point on this is: everything in core should not only "work" but it should also provide a good user experience.

The EOL issue is something that does not seem specific to `assert` but indicates for me that we have a issue somewhere else and it is just visible in `assert` because we verify the message output all the time.